sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -0,0 +1,555 @@
+
+  @RpcMethod("eth_chainId")
+  def chainId: Quantity = new Quantity(networkParams.chainId)
+
+  @RpcMethod("eth_getBalance")
+  @RpcOptionalParameters(1)
+  def getBalance(address: Address, tag: String): Quantity = {
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, tag) { (tagStateView, _) =>
+        new Quantity(tagStateView.getBalance(address.toBytes))
+      }
+    }
+  }
+
+  @RpcMethod("eth_getTransactionCount")
+  @RpcOptionalParameters(1)
+  def getTransactionCount(address: Address, tag: String): Quantity = {
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, tag) { (tagStateView, _) =>
+        new Quantity(tagStateView.getNonce(address.toBytes))
+      }
+    }
+  }
+
+  private def getBlockByTag(nodeView: NV, tag: String): (AccountBlock, SidechainBlockInfo) = {
+    val blockId = getBlockIdByTag(nodeView, tag)
+    val block = nodeView.history
+      .getStorageBlockById(blockId)
+      .getOrElse(throw new RpcException(RpcError.fromCode(RpcCode.UnknownBlock, "Invalid block tag parameter.")))
+    val blockInfo = nodeView.history.blockInfoById(blockId)
+    (block, blockInfo)
+  }
+
+  private def getStateViewAtTag[A](nodeView: NV, tag: String)(fun: (AccountStateView, BlockContext) ⇒ A): A = {
+    val (block, blockInfo) = getBlockByTag(nodeView, tag)
+    val blockContext = new BlockContext(
+      block.header,
+      blockInfo.height,
+      TimeToEpochUtils.timeStampToEpochNumber(networkParams, blockInfo.timestamp),
+      blockInfo.withdrawalEpochInfo.epoch
+    )
+    using(nodeView.state.getStateDbViewFromRoot(block.header.stateRoot))(fun(_, blockContext))
+  }
+
+  private def getBlockIdByTag(nodeView: NV, tag: String): ModifierId = {
+    val history = nodeView.history
+    val blockId = tag match {
+      case "earliest" => history.blockIdByHeight(1)
+      case "finalized" | "safe" => throw new RpcException(RpcError.fromCode(RpcCode.UnknownBlock))
+      case "latest" | "pending" | null => history.blockIdByHeight(history.getCurrentHeight)
+      case height => Try.apply(Numeric.decodeQuantity(height).intValueExact()).toOption.flatMap(history.blockIdByHeight)
+    }
+    ModifierId(
+      blockId
+        .getOrElse(throw new RpcException(new RpcError(RpcCode.InvalidParams, "Invalid block tag parameter", null)))
+    )
+  }
+
+  @RpcMethod("net_version")
+  def version: String = String.valueOf(networkParams.chainId)
+
+  @RpcMethod("eth_gasPrice")
+  def gasPrice: Quantity = {
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, "latest") { (_, blockContext) => new Quantity(blockContext.baseFee) }
+    }
+  }
+
+  private def getTransactionAndReceipt(transactionHash: Hash) = {
+    applyOnAccountView { nodeView =>
+      using(nodeView.state.getView) { stateView =>
+        stateView
+          .getTransactionReceipt(transactionHash.toBytes)
+          .flatMap(receipt => {
+            nodeView.history
+              .blockIdByHeight(receipt.blockNumber)
```

Comment:
Currently this is unnecessarily complex and feels very much backwards: To find a transaction by hash we have to retrieve the receipt first, get the block number from the receipt, retrieve the block and then find the transaction in that block. We could instead build a map from TX hash to TX.